client/src/AppContainer.tsx

```diff
@@ -0,0 +1,61 @@
+import React, { useState, useEffect } from 'react';
+import Keycloak from 'keycloak-js';
+import { KeycloakProvider } from '@react-keycloak/web';
+import { ApolloOfflineClient } from 'offix-client';
+import { ApolloOfflineProvider } from 'react-offix-hooks';
+import { ApolloProvider } from '@apollo/react-hooks';
+import { AppContext } from './AppContext';
+import { keycloakConfig, clientConfig } from './config';
+import { keycloakEnabled, onKeycloakTokens } from './helpers/keycloakHelpers';
+import { Loading } from './components/Loading';
+import { IContainerProps } from './declarations';
+
+const client = new ApolloOfflineClient(clientConfig);
+
+export const AppContainer: React.FC<IContainerProps> = ({ app: App }) => {
+
+  const [initialized, setInitialized] = useState(false);
+
+  // Initialize the client
+  useEffect(() => {
+    client.init().then(() => setInitialized(true));
+  }, []);
+
+  if (!initialized) return <Loading loading={!initialized} />;
+
+  // check if keycloak is configured
+  if (keycloakEnabled(keycloakConfig)) {
+
+    // get keycloak config options
+    const { auth, initConfig} = keycloakConfig;
+
+    // instantiate keycloak
+    const keycloak: Keycloak.KeycloakInstance = new (Keycloak as any)(auth);
```

Comment:
Yes. Keycloak typings are not the best. I will wrap this 3 statements into some helper that creates singleton as part of the keycloak helper. Keycloak instance should be generally singleton that could be required directly - I can be created on root level. The faster it will be created then better because if we render that after init we are duplicating amount of work as page will be refreshed to login and then it will need to be rerendered again. 

Good practice was to keep instance on root so redirect will be instant without rendering any of the react components or initializing the client